File: LeetCode/algorithms/easy/leetcode_1512_Number_of_Good_Pairs/leetcode_1512.py
class Solution:
    # big O(n)
    def numIdenticalPairs(self, nums: List[int]) -> int:
        count = {}
        res = 0
        for num in nums:
            if num in count:
                res += count[num]
            count[num] = count.get(num, 0) + 1
        return res

    # big O(n)
    def solution_II(self, nums: List[int]) -> int:
        res = 0
        count = [0] * 101
        for a in nums:
            res += count[a]
            count[a] += 1
        return res

    # big O(n^2)
    def solution_I(self, nums: List[int]) -> int:
        pairs = []
        for i in range(len(nums)):
            for j in range(i + 1, len(nums)):
                if nums[i] == nums[j]:
                    pairs.append([nums[i], nums[j]])
        return len(pairs)

File: leetCode/problems/easy/leetcode_1512_Number_of_Good_Pairs/leetcode_1512.js
/**
 * @param {number[]} nums
 * @return {number}
 */
var numIdenticalPairs = function (nums) {
  const count = {};
  let res = 0;
  for (const num of nums) {
    if (num in count) {
      res += count[num];
    }
    count[num] = (count[num] || 0) + 1;
  }
  return res;
};

// big O(n)
function solution_II(nums) {
  let res = 0;
  const count = new Array(101).fill(0);
  for (const a of nums) {
    res += count[a];
    count[a] += 1;
  }
  return res;
}

// big O(n^2)
function solution_I(nums) {
  const pairs = [];
  for (let i = 0; i < nums.length; i++) {
    for (let j = i + 1; j < nums.length; j++) {
      if (nums[i] === nums[j]) {
        pairs.push([nums[i], nums[j]]);
      }
    }
  }
  return pairs.length;
}

const nums1 = [1, 2, 3, 1, 1, 3];
console.log(numIdenticalPairs(nums1)); // Output: 4

const nums2 = [1, 1, 1, 1];
console.log(numIdenticalPairs(nums2)); // Output: 6

const nums3 = [1, 2, 3];
console.log(numIdenticalPairs(nums3)); // Output: 0
